src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/CodeActions/Razor/PromoteUsingCodeActionProvider.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Razor.Extensions;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Components;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.CodeActions.Models;
+using Microsoft.CodeAnalysis.Razor.CodeActions.Razor;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Razor.CodeActions;
+
+internal class PromoteUsingCodeActionProvider : IRazorCodeActionProvider
+{
+    public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    {
+        if (context.HasSelection)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var owner = syntaxTree.Root.FindNode(TextSpan.FromBounds(context.StartAbsoluteIndex, context.EndAbsoluteIndex));
+        if (owner is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (owner.FirstAncestorOrSelf<RazorDirectiveSyntax>() is not { } directive ||
+            !directive.IsUsingDirective(out _))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var importFileName = FileKinds.IsLegacy(context.DocumentSnapshot.FileKind)
+                ? MvcImportProjectFeature.ImportsFileName
+                : ComponentMetadata.ImportsFileName;
+
+        var line = context.CodeDocument.Source.Text.Lines.GetLineFromPosition(context.StartAbsoluteIndex);
+        var data = new PromoteToUsingCodeActionParams
+        {
+            ImportsFileName = importFileName,
+            UsingDirective = directive.GetContent(),
```

Comment:
💡 given we have the start and end for the remove, can we avoid passing the content and instead get it from the text on the resolver side?

Reply:
Would have to save another set of start/end, but I guess it would probably still be a bit of a saving most of the time, if the property names are short. I'm not sure how far we really should be going as far as saving JSON serialization with these sorts of things.

Reply:
I think my assumption was that content going into the new file is exactly the same as being removed. That may be where assumption falls apart?

Reply:
It's not because of leading and trailing whitespace, so there is an argument its being a little over-designed.

Actually I should probably add a test for that explicitly, but I was thinking about the case of an `@using` indented inside a Html element or similar.